gui/src/vmm/hw/console/context.rs

```diff
@@ -44,30 +44,29 @@ impl<'a, H: Hypervisor, C: Cpu> DeviceContext<C> for Context<'a, H> {
             let len = self.msg_len.take().ok_or(ExecError::InvalidSequence)?;
             let data = read_ptr(exit, len, self.hv).map_err(|e| ExecError::ReadFailed(off, e))?;
 
-            self.msg
-                .extend_from_slice(unsafe { std::slice::from_raw_parts(data.as_ptr(), len.get()) });
+            self.msg.extend_from_slice(data.as_ref());
         } else if off == offset_of!(ConsoleMemory, commit) {
             // Check if state valid.
             if self.msg_len.is_some() || self.msg.is_empty() {
                 return Err(Box::new(ExecError::InvalidSequence));
-            } else if std::str::from_utf8(&self.msg).is_err() {
-                return Err(Box::new(ExecError::InvalidMsg));
             }
 
+            // Check if valid UTF-8.
+            let _ = std::str::from_utf8(&self.msg).map_err(ExecError::InvalidMsg)?;
+
             // Parse data.
             let commit = read_u8(exit).map_err(|e| ExecError::ReadFailed(off, e))?;
             let ty: ConsoleType = commit
                 .try_into()
-                .map_err(|_| Box::new(ExecError::InvalidCommit(commit)))?;
+                .map_err(|_| ExecError::InvalidCommit(commit))?;
 
             // Trigger event.
             let msg = std::mem::take(&mut self.msg);
 
-            (self.dev.event)(VmmEvent::Log {
-                ty: ty.into(),
-                data: msg.as_ptr().cast(),
-                len: msg.len(),
-            });
+            // Safety: We check the message is valid UTF-8 above.
+            let msg = unsafe { String::from_utf8_unchecked(msg) };
```

Comment:
Why not use `to_owned` on the returned value of `from_utf8` above?

Reply:
Because that would clone the whole vector when we don't have to.

Reply:
I did it like this because it preserves the logic of the original code - check utf8 (and error if not utf8), check commit, mem::take the Vec, and then construct the VmmEvent infallibly.
String::from_utf8 takes and owned vec, meaning that we have to mem::take self.msg and then check if the string is utf8 and it it isn't, put the Vec back into self.msg. Or maybe you're fine with not putting it back. I don't know.
What exactly do you want?

Reply:
Every errors here will exit the VMM immediately so we don't need to put it back.